README.md
# maplibre-basemaps
### Maplibre plugin to add basemaps layer switcher

Check the demo here [Link](https://kaditya97.github.io/maplibre-basemaps/)

## Install

```bash
npm install maplibre-gl maplibre-basemaps
```

or

```html
<link href="https://unpkg.com/maplibre-gl@2.2.0/dist/maplibre-gl.css" rel="stylesheet" />
<script src="https://unpkg.com/maplibre-gl@2.2.0/dist/maplibre-gl.js"></script>
<script src="https://unpkg.com/maplibre-basemaps@0.0.5/dist/maplibre-basemaps.js"></script>
```

## Usage
```jsx
import Maplibre from 'maplibre-gl';
import BasemapControl from 'maplibre-basemaps';
```

## Example usage
Check docs/index.html for example implementation.
```Javascript
const osm = {
    name: "Open Street Map",
    tiles: ['https://tile.openstreetmap.org/{z}/{x}/{y}.png'],
    maxzoom: 18,
    attribution: 'osm'
}
const osmHot = {
    name: "OSM HOT",
    tiles: ['https://a.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png'],
}
const osmCycle = {
    name: "OSM Cycle",
    tiles: ['https://a.tile-cyclosm.openstreetmap.fr/cyclosm/{z}/{x}/{y}.png'],
}
const esriTerrain = {
    name: "Esri Terrain",
    tiles: ['https://server.arcgisonline.com/ArcGIS/rest/services/World_Terrain_Base/MapServer/tile/{z}/{y}/{x}'],
    maxzoom: 13,
    attribution: 'Tiles &copy; Esri &mdash; Source: USGS, Esri, TANA, DeLorme, and NPS',
}
const baseLayers = {
    osm,
    osmHot,
    osmCycle,
    esriTerrain,
}
const basemapControl = new BasemapControl({ basemaps: baseLayers, initialBasemap: "osmHot"  });
map.addControl(basemapControl, 'top-right');
```

### Remove

```Javascript
map.removeControl(basemapControl);
```

### Options for use

- `basemaps` - BaseLayerConfig, base layers objects, **required**
- `initialBasemap` - string, basemap key to visualize, **default: first basemap**
- `width` - string, width for basemaps layers container, **default: '150px'**
- `height` - string, height for basemaps layers container, **default: '100px'**

Layer Configuration
- `name` - string, name to visualize layer in container, **required**
- `tiles` - string[], urls of basemap layer, **required**
- `attribution` - string, attribution for basemap layer, **optional**
- `minZoom` - number, min zoom to display the grid, **default: 0**
- `maxZoom` - number, max zoom to display the grid , **default: 20**



CHANGELOG.md
# Changelog

## 0.0.1

- initial release

## 0.0.2

- minor fixes

## 0.0.3

- typescript

## 0.0.4

- type fix

## 0.0.5

- initalBasempa added instead of visibility option in each layer


src/basemaps.js
/** @typedef {import('maplibre-gl').Map} Map */
/** @typedef {import('./basemaps').BasemapsConfig} BasemapsConfig */

export class Basemaps {
    /**
   * @param {BasemapsConfig} config
   */
    constructor(config) {
        this.basemaps = config.basemaps ?? {};
        this.config = config;
    }

    /**
     * @param {Map} map
     * @returns {HTMLElement}
     */
    onAdd(map) {
        const div = document.createElement('div');
        div.className = 'maplibregl-ctrl maplibregl-ctrl-group maplibre-ctrl-basemap';
        div.innerHTML = `<button>
            <svg fill="#000000" height="18" width="18" version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" 
            viewBox="0 0 421.59 421.59" xml:space="preserve">
            <g>
                <g>
                    <path d="M400.491,291.098l-58.865-36.976l58.864-36.971c2.185-1.372,3.511-3.771,3.511-6.351s-1.326-4.979-3.511-6.352
                        l-58.865-36.977l58.862-36.973c2.185-1.373,3.511-3.771,3.511-6.351s-1.326-4.979-3.511-6.351L214.783,1.149
                        c-2.438-1.532-5.54-1.532-7.979,0L21.1,117.796c-2.185,1.373-3.511,3.771-3.511,6.351c0,2.58,1.326,4.979,3.511,6.351
                        l58.861,36.972l-58.859,36.978c-2.185,1.373-3.51,3.771-3.51,6.351c0,2.58,1.326,4.979,3.511,6.351l58.859,36.97l-58.859,36.979
                        c-2.185,1.372-3.51,3.771-3.51,6.351c0,2.58,1.326,4.979,3.511,6.351l185.7,116.64c1.22,0.766,2.604,1.149,3.989,1.149
                        s2.77-0.383,3.989-1.149L400.491,303.8c2.185-1.372,3.511-3.771,3.511-6.351C404.002,294.869,402.676,292.47,400.491,291.098z
                        M39.189,124.147l171.605-107.79l171.604,107.79l-171.604,107.79L39.189,124.147z M39.191,210.798l54.869-34.471l112.744,70.818
                        c1.219,0.766,2.604,1.149,3.989,1.149c1.385,0,2.77-0.383,3.989-1.149l112.742-70.817l54.875,34.47L210.792,318.582
                        L39.191,210.798z M210.792,405.232L39.191,297.448l54.87-34.472l112.742,70.814c1.22,0.766,2.604,1.149,3.989,1.149
                        s2.77-0.383,3.989-1.149l112.744-70.812l54.876,34.47L210.792,405.232z"/>
                </g>
            </g>
            </svg>
        </button>`;
        const radioContainer = document.createElement('div');
        radioContainer.style.width = this.config.width ?? '150px';
        radioContainer.style.height = this.config.height ?? '100px';
        radioContainer.style.maxWidth = '250px';
        radioContainer.style.maxHeight = '250px';
        radioContainer.style.display = 'none';
        radioContainer.style.position = 'absolute';
        radioContainer.style.top = '0';
        radioContainer.style.left = '0';
        radioContainer.style.backgroundColor = 'white';
        radioContainer.style.border = '2px solid #ccc';
        radioContainer.style.borderRadius = '5px';
        radioContainer.style.zIndex = '100';
        radioContainer.style.flexDirection = 'column';
        radioContainer.style.padding = '2px 0px';
        radioContainer.style.overflowY = 'auto';
        radioContainer.style.overflowX = 'hidden';

        /**
         * @type {string}
         */
        let previouslayer = '';

        /**
         * 
         * @param {*} event 
         */
        const handleChange = (event) => {
            map.setLayoutProperty(previouslayer, 'visibility', 'none');
            map.setLayoutProperty(event.target.value, 'visibility', 'visible');
            previouslayer = event.target.value;
        }

        Object.entries(this.basemaps).map(([key, value]) => {
            const radioOptions = document.createElement('div');
            radioOptions.style.display = 'flex';
            radioOptions.style.alignItems = 'center';
            radioOptions.style.padding = '2px 5px';
            radioOptions.style.cursor = 'pointer';
            radioOptions.onmouseenter = () => {
                radioOptions.style.backgroundColor = '#eee';
            }
            radioOptions.onmouseleave = () => {
                radioOptions.style.backgroundColor = 'white';
            }
            const radio = document.createElement('input');
            radio.style.cursor = 'pointer';
            radio.style.margin = '0px';
            radio.type = 'radio';
            radio.id = `basemap-${key}`;
            radio.name = 'options';
            radio.value = key;
            if (value.visibility === 'visible') {
                previouslayer = key;
                radio.checked = true;
            }
            radio.addEventListener('change', handleChange)
            const label = document.createElement('label');
            label.style.padding = '0px 5px';
            label.style.cursor = 'pointer';
            label.style.display = 'inline-block';
            label.style.width = 'calc(100% - 13px)';
            label.htmlFor = `basemap-${key}`;
            label.textContent = value.name;
            radioOptions.appendChild(radio);
            radioOptions.appendChild(label);
            radioContainer.appendChild(radioOptions);
            div.appendChild(radioContainer);
        });

        div.onmouseenter = () => {
            const basemapControl = document.getElementsByClassName('maplibre-ctrl-basemap')[0];
            const basemaCtrlPosition = basemapControl.getBoundingClientRect();
            const ctrlTop = basemaCtrlPosition.top;
            const ctrlLeft = basemaCtrlPosition.left;

            const mapContainer = map.getCanvas().getBoundingClientRect();
            const mapHeight = mapContainer.height;
            const mapWidth = mapContainer.width;

            radioContainer.style.display = 'flex';

            const radioContainerPosition = radioContainer.getBoundingClientRect();
            const radioHeight = radioContainerPosition.height;
            const radioWidth = radioContainerPosition.width;
            
            if ((ctrlTop + radioHeight) > mapHeight) {
                radioContainer.style.marginTop = `-${radioHeight - basemaCtrlPosition.height}px`;
            } else {
                radioContainer.style.marginTop = '0px';
            }
            if ((ctrlLeft + radioWidth) > mapWidth) {
                radioContainer.style.marginLeft = `-${radioWidth}px`;
            } else {
                radioContainer.style.marginLeft = `${basemaCtrlPosition.width}px`;
            }
        }
        div.onmouseleave = () => {
            radioContainer.style.display = 'none';
        }

        Object.entries(this.basemaps).map(([key, value]) => {
            /**
             * @type {maplibregl.SourceSpecification}
             */
            const source = {
                type: 'raster',
                tiles: value.tiles,
                tileSize: 256,
                minzoom: value.minzoom ?? 0,
                maxzoom: value.maxzoom ?? 18,
                attribution: value.attribution ?? '',
            }

            /**
             * @type {maplibregl.LayerSpecification}
             */
            const layer = {
                id: key,
                type: 'raster',
                source: key,
                layout: {
                    visibility: value.visibility,
                },
            }
            map.addSource(key, source);
            map.addLayer(layer)

        });

        return div;
    }

    onRemove() {
    }
}


rollup.config.js
import pkg from './package.json';
import resolve from '@rollup/plugin-node-resolve';
import typescript from '@rollup/plugin-typescript';
import babel from '@rollup/plugin-babel';
import { terser } from 'rollup-plugin-terser';
import visualizer from 'rollup-plugin-visualizer';

function bundle(filename, options = {}) {
  return {
    input: 'src/index.ts',
    output: {
      file: filename,
      format: 'umd',
      name: 'BasemapControl',
      sourcemap: true
    },
    external: [
      // ...Object.keys(pkg.peerDependencies),
      'fs',
      'path',
    ],
    plugins: [
      resolve(),
      typescript(
        {
          tsconfig: 'tsconfig.json',
          declaration: true,
          declarationDir: 'dist',
        }
      ),
      babel({ babelHelpers: 'runtime' }),
      options.minimize ? terser() : false,
      options.stats ? visualizer({
        filename: filename + '.stats.html',
      }) : false,
    ],
  };
}

export default [
  bundle(pkg.browser.replace('.min', ''), { stats: true }),
  bundle(pkg.browser, { minimize: true }),
];


src/maplibre-basemaps.ts
import { LayerSpecification, SourceSpecification, Map } from 'maplibre-gl';

export interface BaseLayerConfig {
    name: string;
    tiles: Array<string>;
    attribution?: string;
    minZoom?: number;
    maxZoom?: number;
  }

export interface BasemapsConfig {
    basemaps: { [key: string]: BaseLayerConfig };
    initialBasemap?: string;
    width?: string;
    height?: string;
  }

export class BasemapControl {
    basemaps: { [key: string]: BaseLayerConfig };
    initialBasemap?: string;
    config: BasemapsConfig;
    _container: HTMLElement;
    constructor(config: BasemapsConfig) {
        this.basemaps = config.basemaps ?? {};
        this.initialBasemap = config.initialBasemap ?? Object.keys(this.basemaps)[0];
        this.config = config;
        this._container = document.createElement('div');
    }

    onAdd(map: Map): HTMLElement { 
        const div = this._container;
        div.className = 'maplibregl-ctrl maplibregl-ctrl-group maplibre-ctrl-basemap';
        div.innerHTML = `<button>
            <svg fill="#000000" height="18" width="18" version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" 
            viewBox="0 0 421.59 421.59" xml:space="preserve">
            <g>
                <g>
                    <path d="M400.491,291.098l-58.865-36.976l58.864-36.971c2.185-1.372,3.511-3.771,3.511-6.351s-1.326-4.979-3.511-6.352
                        l-58.865-36.977l58.862-36.973c2.185-1.373,3.511-3.771,3.511-6.351s-1.326-4.979-3.511-6.351L214.783,1.149
                        c-2.438-1.532-5.54-1.532-7.979,0L21.1,117.796c-2.185,1.373-3.511,3.771-3.511,6.351c0,2.58,1.326,4.979,3.511,6.351
                        l58.861,36.972l-58.859,36.978c-2.185,1.373-3.51,3.771-3.51,6.351c0,2.58,1.326,4.979,3.511,6.351l58.859,36.97l-58.859,36.979
                        c-2.185,1.372-3.51,3.771-3.51,6.351c0,2.58,1.326,4.979,3.511,6.351l185.7,116.64c1.22,0.766,2.604,1.149,3.989,1.149
                        s2.77-0.383,3.989-1.149L400.491,303.8c2.185-1.372,3.511-3.771,3.511-6.351C404.002,294.869,402.676,292.47,400.491,291.098z
                        M39.189,124.147l171.605-107.79l171.604,107.79l-171.604,107.79L39.189,124.147z M39.191,210.798l54.869-34.471l112.744,70.818
                        c1.219,0.766,2.604,1.149,3.989,1.149c1.385,0,2.77-0.383,3.989-1.149l112.742-70.817l54.875,34.47L210.792,318.582
                        L39.191,210.798z M210.792,405.232L39.191,297.448l54.87-34.472l112.742,70.814c1.22,0.766,2.604,1.149,3.989,1.149
                        s2.77-0.383,3.989-1.149l112.744-70.812l54.876,34.47L210.792,405.232z"/>
                </g>
            </g>
            </svg>
        </button>`;
        const radioContainer = document.createElement('div');
        radioContainer.style.width = this.config.width ?? '150px';
        radioContainer.style.height = this.config.height ?? '100px';
        radioContainer.style.maxWidth = '250px';
        radioContainer.style.maxHeight = '250px';
        radioContainer.style.display = 'none';
        radioContainer.style.position = 'absolute';
        radioContainer.style.top = '0';
        radioContainer.style.left = '0';
        radioContainer.style.backgroundColor = 'white';
        radioContainer.style.border = '2px solid #ccc';
        radioContainer.style.borderRadius = '5px';
        radioContainer.style.zIndex = '100';
        radioContainer.style.flexDirection = 'column';
        radioContainer.style.padding = '2px 0px';
        radioContainer.style.overflowY = 'auto';
        radioContainer.style.overflowX = 'hidden';

        let previouslayer: string = '';

        const handleChange = (event: any) => {
            map.setLayoutProperty(previouslayer, 'visibility', 'none');
            map.setLayoutProperty(event.target.value, 'visibility', 'visible');
            previouslayer = event.target.value;
        }

        Object.entries(this.basemaps).map(([key, value]) => {
            let visibilityValue: string;
            if (this.initialBasemap === key) {
                visibilityValue = 'visible';
            }else {
                visibilityValue = 'none';
            }
            const radioOptions = document.createElement('div');
            radioOptions.style.display = 'flex';
            radioOptions.style.alignItems = 'center';
            radioOptions.style.padding = '2px 5px';
            radioOptions.style.cursor = 'pointer';
            radioOptions.onmouseenter = () => {
                radioOptions.style.backgroundColor = '#eee';
            }
            radioOptions.onmouseleave = () => {
                radioOptions.style.backgroundColor = 'white';
            }
            const radio = document.createElement('input');
            radio.style.cursor = 'pointer';
            radio.style.margin = '0px';
            radio.type = 'radio';
            radio.id = `basemap-${key}`;
            radio.name = 'options';
            radio.value = key;
            if (visibilityValue === 'visible') {
                previouslayer = key;
                radio.checked = true;
            }
            radio.addEventListener('change', handleChange)
            const label = document.createElement('label');
            label.style.padding = '0px 5px';
            label.style.cursor = 'pointer';
            label.style.display = 'inline-block';
            label.style.width = 'calc(100% - 13px)';
            label.htmlFor = `basemap-${key}`;
            label.textContent = value.name;
            radioOptions.appendChild(radio);
            radioOptions.appendChild(label);
            radioContainer.appendChild(radioOptions);
            div.appendChild(radioContainer);
        });

        div.onmouseenter = () => {
            const basemapControl = document.getElementsByClassName('maplibre-ctrl-basemap')[0];
            const basemaCtrlPosition = basemapControl.getBoundingClientRect();
            const ctrlTop = basemaCtrlPosition.top;
            const ctrlLeft = basemaCtrlPosition.left;

            const mapContainer = map.getCanvas().getBoundingClientRect();
            const mapHeight = mapContainer.height;
            const mapWidth = mapContainer.width;

            radioContainer.style.display = 'flex';

            const radioContainerPosition = radioContainer.getBoundingClientRect();
            const radioHeight = radioContainerPosition.height;
            const radioWidth = radioContainerPosition.width;
            
            if ((ctrlTop + radioHeight) > mapHeight) {
                radioContainer.style.marginTop = `-${radioHeight - basemaCtrlPosition.height}px`;
            } else {
                radioContainer.style.marginTop = '0px';
            }
            if ((ctrlLeft + radioWidth) > mapWidth) {
                radioContainer.style.marginLeft = `-${radioWidth}px`;
            } else {
                radioContainer.style.marginLeft = `${basemaCtrlPosition.width}px`;
            }
        }
        div.onmouseleave = () => {
            radioContainer.style.display = 'none';
        }

        Object.entries(this.basemaps).map(([key, value]) => {
            let visibilityValue: string;
            if (this.initialBasemap === key) {
                visibilityValue = 'visible';
            }else {
                visibilityValue = 'none';
            }
            const source: SourceSpecification = {
                type: 'raster',
                tiles: value.tiles,
                tileSize: 256,
                minzoom: value.minZoom ?? 0,
                maxzoom: value.maxZoom ?? 18,
                attribution: value.attribution ?? '',
            }

            const layer: LayerSpecification = {
                id: key,
                type: 'raster',
                source: key,
                layout: {
                    visibility: visibilityValue as  "visible" | "none",
                },
            }
            map.addSource(key, source);
            map.addLayer(layer)

        });

        return div;
    }

    onRemove(map: Map) {
        this._container.parentNode?.removeChild(this._container);
        Object.entries(this.basemaps).map(([key]) => {
            map.removeLayer(key);
            map.removeSource(key);
        });
    }
}
